internal/server/web_server_e2e_test.go

```diff
@@ -0,0 +1,203 @@
+package server
+
+import (
+	"bytes"
+	"encoding/json"
+	"math/big"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/satsuma-data/node-gateway/internal/config"
+	"github.com/satsuma-data/node-gateway/internal/jsonrpc"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestServeHTTP_ForwardsToSoleHealthyUpstream(t *testing.T) {
```

Comment:
These kind of remind me of a halfway point between unit tests and an integration tests!

Reply:
Yep, they don't bring up a full set of infrastructure (real ETH nodes, etc.) so they're not 100% realistic, but it's super convenient that they're all written in the same language (i.e. more control, better tooling) and they're also reasonably efficient to run, unlike a full-blown setup.